core/src/main/java/org/apache/iceberg/DataFiles.java

```diff
@@ -248,6 +268,16 @@ public Builder withPartitionPath(String newPartitionPath) {
       return this;
     }
 
+    public Builder withPartitionValues(List<String> partitionValues) {
+      Preconditions.checkArgument(
+          isPartitioned || partitionValues.isEmpty(),
+          "Cannot add partition data for an unpartitioned table");
+      if (!partitionValues.isEmpty()) {
```

Comment:
This branch is never false

Reply:
Do you mean `partitionValues.isEmpty()` can never be true logically? Or it is never true because `withPartitionValues` is never called with an empty `partitionValues`?
If the latter, that is because I'm the only one who has written code calling `withPartitionValues` but there is nothing to stop future callers from calling it with empty `partitionValues`.
I copied the code from the existing `withPartitionPath`. Perhaps the `if` in both methods is overkill and protects against calls that don't exist (but in principle could occur).

Reply:
This is actually a little more confusing still, "fillFromValues" throws an error if you have the wrong number of values so currently the logic looks like

1. isPartitioned + partitionValues.isEmpty - doesNothing
2. isPartitioned + parititonValues.size != partitonSpec.size && isNotEmpty - Error
3. isPartitioned + partitionValues.size == partitionSpec.size - does something
4. not partitioned + partitionvalues.isEmpty - doesNothing
5. not partitioned + parittionValues.nonEmpty - Error

In my opinion here everything but 3 should be illegal.

Reply:
I can change the `Preconditions.checkArgument` in `withPartitionValues` to
```
      Preconditions.checkArgument(
          isPartitioned && !partitionValues.isEmpty(),
          "Table must be partitioned and partition values must not be empty");
```
Then 1 and 4 will throw an error.
Only 3 will be allowed.

Reply:
Actually, I see now that we should leave things as they are.
`SparkTableUtil.importUnpartitionedSparkTable` actually does call `TableMigrationUtil.listPartition`, with an empty `Map` for `partition`, and so that will call `DataFiles.Builder#withPartitionValues` with an empty `List` for `partitionValues`. That should be allowed.

(I tried the change I just proposed above, and some tests in `TestMigrateTableProcedure` failed.)

Reply:
So if it's unpartitioned then it must be empty. So either it is partitioned and values is as long as the spec. Or it is unpartitioned and empty?

Reply:
Yes, that is correct.
Right now we allow 1 and 4. We should allow 4, but to be strict, we should not allow 1. So, if we want to be strict, we can use
```
      Preconditions.checkArgument(
          (isPartitioned && !partitionValues.isEmpty())
              || (!isPartitioned && partitionValues.isEmpty()),
          "Table must be partitioned or partition values must be empty");
```
We still need the `if (!partitionValues.isEmpty())` though.

Reply:
Or more succinctly,
```
      Preconditions.checkArgument(
          isPartitioned ^ partitionValues.isEmpty(),
          "Table must be partitioned or partition values must be empty");
```
I didn't realize that `^` could be used as a logical xor as well as bitwise xor.